Add PatternedHexPeripheralModifier and TransitionLayerTools

miaoyinb opened the pull request:
closes #20641

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

miaoyinb committed:
Add PatternedHexPeripheralModifier and TransitionLayerTools

closes #20641

minor doxygen doc fix ref #20641

minor cleanup ref #20641

CIVET fix 1

CIVET fix 2 SQA and distributed-mesh test ref #20641

CIVET fix 3 ref #20641

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/994872/) on d632afd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 71e154c564a08e6b29f64374f094721c700d8141`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit acc498685d1453fda11c2000886fdacdae6ae333
+Subproject commit 5e3373c5383afdaa2dafca6495923348131f0b22
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/995138/) on 391a437 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20645/site)

This comment will be updated on new commits.

miaoyinb committed:
Some functions moved to TransitionLayerTools namespace ref #20641

CIVET fix 4 add header files for non-unity tests ref #20641

CIVET fix 5 add header files for non-unity tests 2 ref #20641

CIVET fix 6 fix an assertion in dbg mode ref #20641

Apply suggestions from code review

Guillaume suggested syntax/doc update ref #20641

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

apply review suggestions on syntax and docs ref #20641

Clang-format fix 2 ref #20641

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/995722/) on 262a693 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7e4878</th>
      <th colspan=4>#20645 262a69</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.55%</td>
      <td>82.60%</td>
      <td>+0.05%</td>
      <td>93.52%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73710</td>
      <td>74114</td>
      <td>+404</td>
      <td>404</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15580</td>
      <td>15608</td>
      <td>+28</td>
      <td>28</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20645/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20645/coverage/framework)


## Modules coverage

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>7e4878</th>
      <th colspan=4>#20645 262a69</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.09%</td>
      <td>95.29%</td>
      <td>+0.20%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2651</td>
      <td>2769</td>
      <td>+118</td>
      <td>122</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>137</td>
      <td>137</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20645/coverage_diff/reactor.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20645/coverage/reactor)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20645/coverage/xfem)

</details>



This comment will be updated on new commits.

miaoyinb commented:
@eshemon @shikhar413 FYI

@loganharbour @GiudGiud Another PR to add peripheral region modification capability to facilitate assembly stitching. I moved the tools used for such modification into framework as it may be useful for others. Thank you in advance for your review!

miaoyinb commented on framework/include/utils/TransitionLayerTools.h:
```diff
@@ -0,0 +1,210 @@
+ * @param bid ID of the boundary to be examined
+ * @return whether the boundary works with the algorithm
+ */
+bool isBoundaryValid(ReplicatedMesh & mesh,
```
These are similar to the methods in `PeripheralRingMeshGenerator`. Will consolidate them in either this PR or #20638

miaoyinb replied:
Done

GiudGiud commented on framework/src/utils/TransitionLayerTools.C:
```diff
@@ -0,0 +1,614 @@
+}
+
+bool
+isBoundaryValid(ReplicatedMesh & mesh,
```
how big is the overlap with PeripheralRingMeshGenerator's version of this

miaoyinb replied:
A lot. My plan is to move it here into the namespace as this is in the framework.

GiudGiud reviewed (changes requested):
only looked at Transition Layer things

miaoyinb commented on framework/doc/content/source/meshgenerators/TransitionLayerConnector.md:
```diff
@@ -0,0 +1,24 @@
+      id=example_tlc
+      caption=A typical example of using `TransitionLayerConnector` to connect two square meshes together.
+
+All the other meshing options are the same as [`TransitionLayerGenerator`](/TransitionLayerGenerator.md).
```
It's only XY for now.

It cannot directly do two concentric circle. But you can do two halfs and stitch.

miaoyinb commented on framework/include/meshgenerators/TransitionLayerGenerator.h:
```diff
@@ -0,0 +1,49 @@
+ * This TransitionLayerGenerator object is designed to generate a transition layer with two
+ * sides containing different node numbers.
+ */
+class TransitionLayerGenerator : public MeshGenerator
```
Any suggestion?

miaoyinb replied:
Use `StandaloneTransitionLayerGenerator` for now.

miaoyinb commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,206 @@
+      "parameter is not used if other biasing option is selected.");
+  params.addParam<bool>(
+      "keep_inputs",
+      false,
```
good point

miaoyinb commented on framework/include/meshgenerators/TransitionLayerConnector.h:
```diff
@@ -0,0 +1,67 @@
+  std::unique_ptr<MeshBase> & _input_1;
+  /// The mesh which contains the second input boundary
+  std::unique_ptr<MeshBase> & _input_2;
+  /// ID of the external boundary of the input mesh 1
+  std::vector<boundary_id_type> _input_mesh_1_external_bids;
+  /// ID of the external boundary of the input mesh 2
+  std::vector<boundary_id_type> _input_mesh_2_external_bids;
```
right, id does not need to be a class member

miaoyinb commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,206 @@
+  params.addRequiredRangeCheckedParam<unsigned int>(
+      "num_layers", "num_layers>0", "Layers of elements for transition.");
+  params.addParam<subdomain_id_type>("block_id", 1, "ID to be assigned to the block.");
+  params.addParam<boundary_id_type>(
+      "input_boundary_1_id",
+      10000,
+      "Boundary ID to be assigned to the boundary defined by positions_vector_1.");
+  params.addParam<boundary_id_type>(
+      "input_boundary_2_id",
+      10000,
+      "Boundary ID to be assigned to the boundary defined by positions_vector_2.");
```
By default, the external boundary is 10000, which can be used for stitching if input meshes are kept.

If the input mesh are not kept, then the transition layer will need to have boundary ids for these sides.

miaoyinb commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,206 @@
+
+  for (unsigned int i = 1; i < _input_mesh_1_external_bids.size(); i++)
+  {
+    MooseMeshUtils::changeBoundaryId(
+        *input_mesh_1, _input_mesh_1_external_bids[i], _input_mesh_1_external_bids.front(), true);
```
If we do not keep the input meshes, it does not matter messing up its boundary info.
If  we keep the input meshes, these boundary will be used in stitching and become interface.
I understand your concern. Maybe:
(1) create a new boundary id and keep all the old ones;
or (2) just do this and make it clear in the docs

Let me know

miaoyinb replied:
Will mention it in the doc for now.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001417/) on c23edc4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c90b1f04c38b5fa8c35d8b16e7ca4ae7c7ad7115`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001417/) on c23edc4 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001417/) on c23edc4 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1001417/) on c23edc4 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

miaoyinb commented:
@GiudGiud Not sure what happened to the Apple Silicon tests. I don't have any M1 Macs.

GiudGiud commented:
Me neither. But the diffs are very significant so something's up. I ll talk to the team Monday see how to get access to one temporarily

GiudGiud commented:
@cticenhour offered to have a look at this. I m sorry I forgot to update here
We unfortunately cannot give you access to the M1 machines we have, they are very our internal build boxes.

miaoyinb commented:
> @cticenhour offered to have a look at this. I m sorry I forgot to update here We unfortunately cannot give you access to the M1 machines we have, they are very our internal build boxes.

Thanks @GiudGiud @cticenhour

cticenhour commented:
@miaoyinb Just to give somewhat of a timeline for you, I am swamped with other Apple Silicon tasks today, so I will try to get to this tomorrow morning.

miaoyinb commented:
> @miaoyinb Just to give somewhat of a timeline for you, I am swamped with other Apple Silicon tasks today, so I will try to get to this tomorrow morning.

Thanks for the heads-up.

miaoyinb commented:
> @miaoyinb Just to give somewhat of a timeline for you, I am swamped with other Apple Silicon tasks today, so I will try to get to this tomorrow morning.

@cticenhour Just a friendly reminder. Let me know once you got a chance to look into this issue. Thanks,

cticenhour commented on framework/src/utils/TransitionLayerTools.C:
```diff
@@ -0,0 +1,614 @@
+#include "libmesh/face_quad4.h"
+
+// C++ includes
+#include <cmath> // provides round, not std::round (see http://www.cplusplus.com/reference/cmath/round/)
```
@miaoyinb Working through this on M1 today, but I had a quick question. Is this include still necessary? It doesn't look like it's used here or in the reactor module.

miaoyinb replied:
Thanks, @cticenhour.
I remember that I added this include for `elementsCreationFromNodesVectorsQuad()` here.

cticenhour replied:
It doesn't look like its used any longer for `elementsCreationFromNodesVectorsQuad`, so we should clean this up. Will add to a more formal review.

miaoyinb committed:
minor casting fix ref #20641

clang-format fix 3 ref #20641

Apply suggestions from code review

Remove unused includes ref #20641

Co-authored-by: Casey Icenhour <cticenho@ncsu.edu>

clang-format fix 4 ref #20641

Apply suggestions from code review

Improve comments/syntax ref #20641

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/tests ref #20641

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1012009/) on ea296aa wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3587c9b242493465a46a69beb02b09ef93da0b01`

cticenhour commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,208 @@
+#include "CastUniquePointer.h"
+#include "libmesh/node.h"
+
+// C++ includes
+#include <cmath>
```
```suggestion
```

Unnecessary, so it should be removed.

cticenhour commented on framework/src/meshgenerators/TransitionLayerGenerator.C:
```diff
@@ -0,0 +1,104 @@
+#include "TransitionLayerGenerator.h"
+#include "TransitionLayerTools.h"
+
+// C++ includes
+#include <cmath>
+#include <iostream>
```
```suggestion
```
Uneccessary

cticenhour commented on framework/src/utils/TransitionLayerTools.C:
```diff
@@ -0,0 +1,614 @@
+#include "libmesh/face_tri3.h"
+#include "libmesh/face_quad4.h"
+
+// C++ includes
+#include <cmath> // provides round, not std::round (see http://www.cplusplus.com/reference/cmath/round/)
```
```suggestion
```

No longer used in the methods in this class, so should be cleaned up.

cticenhour reviewed (changes requested):
`cmath` and `iostream` should be cleaned up since they don't appear to be necessary. Non-unity build with a full test shows that they're no longer being used.

miaoyinb commented on framework/src/utils/TransitionLayerTools.C:
```diff
@@ -0,0 +1,614 @@
+  {
+    // calculate number of nodes in each sublayer
+    node_number_vec.push_back(
+        (unsigned int)(vec_1_node_num + (long)(increment * i + 0.5 - (increment < 0))));
```
Thanks, @cticenhour. Got help from @shikhar413 who has a M1 Mac👍. It was lucky that my first trial worked 😄.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1012422/) on fb321f2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3587c9b242493465a46a69beb02b09ef93da0b01`

GiudGiud commented on framework/include/utils/TransitionLayerTools.h:
```diff
@@ -0,0 +1,210 @@
+                     const boundary_id_type bid);
+
+/**
+ * Decides whether a boundary of a given mesh works is an external boundary.
```
```suggestion
 * Decides whether a boundary of a given mesh is an external boundary.
```

GiudGiud commented on framework/include/utils/TransitionLayerTools.h:
```diff
@@ -0,0 +1,210 @@
+                        const unsigned int boundary_node_num,
+                        const unsigned int i);
+
+/**
+ * Decides whether all the Points of a vector of Points are on the XY plane
+ * @param vec_pts vector of points to be examined
+ * @return whether all the Points are on the XY plane
+ */
+bool isXYPlane(const std::vector<Point> vec_pts);
```
I'd rather we have a isCoPlanar routine and add it to MooseMeshUtils if that isnt too much work

miaoyinb replied:
Will do.

GiudGiud commented on framework/include/utils/TransitionLayerTools.h:
```diff
@@ -0,0 +1,210 @@
+                              const bool quad_elem = false);
+
+/**
+ * Generates a 2D mesh based on a 2D vector of Nodes using QUAD4 elements
```
this seems like a very general capability, but it s a little more specific right? It only works with a certain positioning of the nodes right?

miaoyinb replied:
The nodes need to be on XY Plane for now. Will add this to the comments.

GiudGiud commented on framework/include/utils/TransitionLayerTools.h:
```diff
@@ -0,0 +1,210 @@
+                                          const boundary_id_type end_side_boundary_id);
+
+/**
+ * Generates a 2D mesh based on a 2D vector of Nodes using TRI3 elements
```
same comment

miaoyinb replied:
Same

GiudGiud commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,205 @@
+  params.addRequiredParam<MeshGeneratorName>("input_mesh_2",
+                                             "The input mesh that contains boundary_2");
+  params.addRequiredParam<std::vector<BoundaryName>>(
+      "boundary_1", "the first boundary that need to be connected.");
```
```suggestion
      "boundary_1", "the first boundary that needs to be connected.");
```

GiudGiud commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,205 @@
+  params.addRequiredParam<std::vector<BoundaryName>>(
+      "boundary_1", "the first boundary that need to be connected.");
+  params.addRequiredParam<std::vector<BoundaryName>>(
+      "boundary_2", "the second boundary that need to be connected.");
```
```suggestion
      "boundary_2", "the second boundary that needs to be connected.");
```

GiudGiud commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,205 @@
+  params.addParam<Point>(
+      "mesh_2_shift", Point(0.0, 0.0, 0.0), "The translate vector to be applied to input_mesh_2");
+  params.addRequiredRangeCheckedParam<unsigned int>(
+      "num_layers", "num_layers>0", "Layers of elements for transition.");
```
```suggestion
      "num_layers", "num_layers>0", "Number of layers of elements created between the boundaries.");
```

GiudGiud commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,205 @@
+  auto input_mesh_2 = dynamic_cast<ReplicatedMesh *>(_input_2.get());
+  if (!input_mesh_1)
+    paramError("input_mesh_1", "Input is not a replicated mesh, which is required.");
+  if (*(input_mesh_1->elem_dimensions().begin()) != 2 ||
+      *(input_mesh_1->elem_dimensions().rbegin()) != 2)
```
why not use mesh_i->mesh_dimension() ?

miaoyinb replied:
`mesh_dimension()` is the maximum of `elem_dimensions()`. I used the current one to make sure that all the elements are 2D.

GiudGiud commented on framework/src/meshgenerators/TransitionLayerConnector.C:
```diff
@@ -0,0 +1,205 @@
+    mesh->stitch_meshes(
+        *input_mesh_1, _input_boundary_1_id, input_mesh_1_external_bids.front(), TOLERANCE, true);
+    mesh->stitch_meshes(
+        *input_mesh_2, _input_boundary_2_id, input_mesh_2_external_bids.front(), TOLERANCE, true);
```
```
void libMesh::[ReplicatedMesh](https://libmesh.github.io/doxygen/classlibMesh_1_1ReplicatedMesh.html)::stitch_meshes	(	const ReplicatedMesh & 	other_mesh,
[boundary_id_type](https://libmesh.github.io/doxygen/namespacelibMesh.html#a3f4d8a52b5ef6ff6156ca209e12b5564) 	this_mesh_boundary,
[boundary_id_type](https://libmesh.github.io/doxygen/namespacelibMesh.html#a3f4d8a52b5ef6ff6156ca209e12b5564) 	other_mesh_boundary,
[Real](https://libmesh.github.io/doxygen/namespacelibMesh.html#aef28359fb4851abec08fc99cfbac118d) 	tol = [TOLERANCE](https://libmesh.github.io/doxygen/namespacelibMesh.html#ac935f799caf032c4008c5aa8facebaae),
bool 	clear_stitched_boundary_ids = false,
bool 	verbose = true,
bool 	use_binary_search = true,
bool 	enforce_all_nodes_match_on_boundaries = false 
)
```
I think you want to pass true to the last one too?

miaoyinb replied:
Will do.

GiudGiud commented on modules/reactor/doc/content/source/meshgenerators/PatternedHexPeripheralModifier.md:
```diff
@@ -0,0 +1,88 @@
+
+!syntax description /Mesh/PatternedHexPeripheralModifier
+
+## Overview
+
+This `PatternedHexPeripheralModifier` class utilizes [`TransitionLayerTools`](/TransitionLayerTools.md) to replace the outmost layer of the quad elements of the hexagonal assembly mesh generated by [`PatternedHexMeshGenerator`](PatternedHexMeshGenerator.md) (or its derived class [`HexIDPatternedMeshGenerator`](HexIDPatternedMeshGenerator.md)) with a transition layer consisting of triangular eleme ...
```
this must mention this is in 2D

miaoyinb replied:
`PatternedHexMeshGenerator` itself is 2D. Will add mentioning of 2D anyway.

GiudGiud reviewed (commented):
partial review

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -382,10 +383,38 @@ class PolygonMeshGeneratorBase : public MeshGenerator
    * @param lower_azi lower boundary of the azimuthal angles to be collected
    * @param upper_azi upper boundary of the azimuthal angles to be collected
    * @param return_type whether angle values or tangent values are returned
    * @param calculate_origin whether the mesh origin is calculated based on the centroid position
+   * @param bid id of the boundary of which the nodes' azimuthal angles are collected
```
```suggestion
   * @param bid id of the boundary of which the nodes' azimuthal angles are collected
   * @param calculate_origin whether the mesh origin is calculated based on the centroid position
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/tests:
```diff
@@ -0,0 +1,61 @@
+    input = 'single_hex.i'
+    exodiff = 'single_hex_in.e'
+    cli_args = '--mesh-only'
+    requirement ='The system shall generate a hexagonal assembly mesh with modified peripheral region to faciliate further stitching.'
```
```suggestion
    requirement ='The system shall generate a hexagonal assembly mesh then modify the peripheral region to facilitate further stitching.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/tests:
```diff
@@ -0,0 +1,61 @@
+    input = 'patterned.i'
+    exodiff = 'patterned_in.e'
+    cli_args = '--mesh-only'
+    requirement ='The system shall generate hexagonal assembly meshes with peripheral regions modified and stiche them together to form a core mesh.'
```
```suggestion
    requirement ='The system shall generate hexagonal assembly meshes with peripheral regions modified in order to stitch them together to form a core mesh.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/tests:
```diff
@@ -0,0 +1,61 @@
+    cli_args = 'Mesh/pmg/input=hex
+                Mesh/final_generator=pmg'
+    expect_err = 'The input mesh is generated by a mesh generator that is not compatible with PatternedHexPeripheralModifier.'
+    requirement ='The system shall throw an error if input mesh is generated by an incompatible mesh generator.'
```
```suggestion
    requirement ='The system shall throw an error if the input mesh to the hexagonal mesh boundary modifier is generated by an incompatible mesh generator.'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/tests:
```diff
@@ -0,0 +1,61 @@
+    cli_args = 'Mesh/pmg/extra_id_names_to_modify=assembly_id
+                Mesh/pmg/new_extra_id_values_to_assign=10'
+    expect_err = 'The parameter contains an extra element integer that does not exist in the input mesh.'
+    requirement ='The system shall throw an error if the specified reporting id does not exist on the input mesh.'
```
```suggestion
    requirement = 'The system shall throw an error if the specified reporting id to be modified does not exist on the input mesh.'
```

GiudGiud commented on test/tests/meshgenerators/transition_layer_connector/tests:
```diff
@@ -0,0 +1,52 @@
+    cli_args = '--mesh-only squares_err.e
+               Mesh/tlc/boundary_1="top bottom"'
+    expect_err = 'The provided boundary is not a single-segment boundary.'
+    requirement ="The system shall throw an error if the input boundary has multiple segments."
```
```suggestion
    requirement = "The system shall throw an error if one of the input boundary to generate a mesh in-between has multiple segments."
```

miaoyinb committed:
Add isCoPlanar() ref #20641

Other syntax and comment changes ref #20641

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1018522/) on b16b1aa wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7c2a607483f2db3424434b3c32f51785e200f8d1`

miaoyinb committed:
Improve boundaries to use node match checking during stitching ref #20641

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1018695/) on 8cbc9ff wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8175f8cb0ba17eaa5473470393b2c535986729a3`

miaoyinb committed:
Consolidate isBoundaryValid() ref #20641

clang-format fix ref #20641

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1018728/) on 9676731 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8175f8cb0ba17eaa5473470393b2c535986729a3`

miaoyinb committed:
StandaloneTransitionLayerGenerator rename ref #20641

clang-format fix 5 ref #20641

Apply suggestions from code review

Typos and syntax fix

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1019004/) on 5f8573e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 00878b3d2449ef56b5b5e5ba3e6067324fa72529`

GiudGiud commented on modules/reactor/src/meshgenerators/PatternedHexPeripheralModifier.C:
```diff
@@ -0,0 +1,262 @@
+                                      ->id());
+    }
+  }
+  // This eliminate the duplicate elements, which happen at the vertices.
```
```suggestion
  // This eliminates the duplicate elements, which happen at the vertices.
```

GiudGiud commented on modules/reactor/src/meshgenerators/PatternedHexPeripheralModifier.C:
```diff
@@ -0,0 +1,262 @@
+    }
+  }
+
+  // Refresh the mesh as it lost some stuffs.
```
```suggestion
  // Refresh the mesh as it lost some elements
```

GiudGiud commented on modules/reactor/include/meshgenerators/PatternedHexPeripheralModifier.h:
```diff
@@ -0,0 +1,65 @@
+  boundary_id_type _input_mesh_external_bid;
+
+  /**
+   * Assign extra element intergers to the newly generated transition layer mesh based on the
```
```suggestion
   * Assign extra element integers to the newly generated transition layer mesh based on the
```

GiudGiud commented on modules/reactor/include/meshgenerators/PatternedHexPeripheralModifier.h:
```diff
@@ -0,0 +1,65 @@
+   * Assign extra element intergers to the newly generated transition layer mesh based on the
+   * nearest element of the deleted original mesh
+   * @param mesh the mesh that will be modified to add extra element integer values
+   * @param ref_extra_ids a data structure containing the extra element interger information of the
```
```suggestion
   * @param ref_extra_ids a data structure containing the extra element integer information of the
```

GiudGiud commented on modules/reactor/src/meshgenerators/PatternedHexPeripheralModifier.C:
```diff
@@ -0,0 +1,262 @@
+  // Sorry it's O(MN) :-(
+  for (const auto & elem : as_range(mesh.active_elements_begin(), mesh.active_elements_end()))
+  {
+    std::vector<Real> tmp_dist;
+    for (auto & pt_extra_id : ref_extra_ids)
+    {
+      tmp_dist.push_back((pt_extra_id.first - elem->true_centroid()).norm());
+    }
```
resize it first it ll be cheaper

GiudGiud replied:
also pre-compute the centroid. true_centroid costs a lot

GiudGiud commented on modules/reactor/src/meshgenerators/PatternedHexPeripheralModifier.C:
```diff
@@ -0,0 +1,262 @@
+}
+
+void
+PatternedHexPeripheralModifier::addExtraElemIntegers(
```
let's call this "transferExtraElemIntegers"

miaoyinb commented:
Thanks @GiudGiud. BTW, what happened to the CIVET?

GiudGiud commented:
what do you mean what happened to civet? is there an unexpected test failure?

miaoyinb commented:
> what do you mean what happened to civet? is there an unexpected test failure?

It did not run for a while. It seems to be fixed by a new commit.

GiudGiud commented:
No idea. It can get brain freezes when we reset the machines. Usually the people doing that maintenance reset all the jobs but maybe they missed that one

miaoyinb committed:
transferExtraElemIntegers() renaming and optimization ref #20641

miaoyinb commented on modules/reactor/src/meshgenerators/PatternedHexPeripheralModifier.C:
```diff
@@ -0,0 +1,262 @@
+        isParamValid("new_extra_id_values_to_assign")
+            ? getParam<std::vector<dof_id_type>>("new_extra_id_values_to_assign")
+            : std::vector<dof_id_type>()),
+    _pattern_pitch_meta(declareMeshProperty<Real>("pattern_pitch_meta", 0.0)),
```
Yes.

miaoyinb commented on modules/reactor/src/meshgenerators/PatternedHexPeripheralModifier.C:
```diff
@@ -0,0 +1,262 @@
+            ? getParam<std::vector<dof_id_type>>("new_extra_id_values_to_assign")
+            : std::vector<dof_id_type>()),
+    _pattern_pitch_meta(declareMeshProperty<Real>("pattern_pitch_meta", 0.0)),
+    // Use HexagonConcentricCircleAdaptiveBoundaryMeshGenerator for control drum meshing
+    _is_control_drum_meta(declareMeshProperty<bool>("is_control_drum_meta", false)),
```
> so you're having to propagate this info for every modification of the mesh part?

Yes. The metadata needs to be propagated so that the mesh can be used by `PatternedHexMG`

miaoyinb replied:
> and what if we re currently wrapping a control drum? Shouldnt we be propagating that metadata?
User will need to use `HexagonConcentricCircleAdaptiveBoundaryMeshGenerator` for control drum. That MG adapters neighboring  meshes so that peripheral modification is not needed.

miaoyinb committed:
exceptions for isBoundaryValid()  ref #20641

clang-format fix 6 #20641

Apply suggestions from code review

Remove extra line

miaoyinb commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -430,9 +431,37 @@ class PolygonMeshGeneratorBase : public MeshGenerator
    * @param upper_azi upper boundary of the azimuthal angles to be collected
    * @param return_type whether angle values or tangent values are returned
+   * @param bid id of the boundary of which the nodes' azimuthal angles are collected
+   * @param calculate_origin whether the mesh origin is calculated based on the centroid position
+   * @param input_origin_x precalculated mesh origin coordinate x
+   * @param input_origin_y precalculated mesh origin coordinate y
```
Yes. `input_origin_x` and `input_origin_y` are not used by default. But if `calculate_origin` is set as false, the default provided `input_origin_x` and `input_origin_y` may be used.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1022297/) on fcff7e1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ae7bbf9dd95ed4d78ac76101f9db7fe505059756`

miaoyinb commented on framework/src/utils/TransitionLayerTools.C:
```diff
@@ -0,0 +1,621 @@
+                const boundary_id_type bid)
+{
+  max_node_radius = 0.0;
+  // invalid_type = 0;
```
```suggestion
```

GiudGiud commented:
if you could reduce the commit history for this PR too please. Same as the other, squashing the minor commits

miaoyinb commented:
@GiudGiud Let me know if any additional revision is needed. Thanks.

miaoyinb committed:
CIVET libmesh test fix ref #20641

CIVET libmesh test fix 2 ref #20641

Apply suggestions from code review

Minor revision suggestions

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud commented on framework/src/utils/TransitionLayerTools.C:
```diff
@@ -0,0 +1,623 @@
+    std::sort(ordered_node_azi_list.begin(), ordered_node_azi_list.end());
+    if (ordered_node_azi_list.front() * ordered_node_azi_list.back() < 0.0)
+    {
+      // This is invalid type #3
+      throw(unsigned short) 3;
```
use real exceptions. see the throw/catch MooseExceptions  in the framework for example

GiudGiud reviewed (changes requested):
so this is getting close except for the name. It's a transition layer in the context of meshing around assemblies in a reactor, but in a general framework context it's not really. 

I went around the office to brainstorm a name, we converged on 
FillBetweenSidesetsGenerator for the main object.

This is still too general in my opinion but others like it. 
Imo this is more a FillBetweenTwoSetsofContiguous1DSidesetsGenerator

so if you could please find a more descriptive / less-wide scope name we should be able to get this in. 
Same applies to TransitionLayerTools and StandaloneTransitionLayerGenerator

miaoyinb commented:
> so this is getting close except for the name. It's a transition layer in the context of meshing around assemblies in a reactor, but in a general framework context it's not really.
> 
> I went around the office to brainstorm a name, we converged on FillBetweenSidesetsGenerator for the main object.
> 
> This is still too general in my opinion but others like it. Imo this is more a FillBetweenTwoSetsofContiguous1DSidesetsGenerator
> 
> so if you could please find a more descriptive / less-wide scope name we should be able to get this in. Same applies to TransitionLayerTools and StandaloneTransitionLayerGenerator

Okay. I think we can call them `FillBetweenSidesetsGenerator`, `FillBetweenPointVectorsGenerator`, and `FillBetweenPointVectorsTools`.

miaoyinb committed:
rename classes ref #20641

rename classes 2 ref #20641

MooseExceptions ref #20641

CIVET clang-format fix

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1030361/) on 93c21ad wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e042275cde8355ad50389dc2e69963d0aec62ae7`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1030361/) on 93c21ad wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1030361/) on 93c21ad wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1030361/) on 93c21ad wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1030395/) on 53e8498 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1030395/) on 53e8498 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1030395/) on 53e8498 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

GiudGiud commented on framework/src/meshgenerators/FillBetweenSidesetsGenerator.C:
```diff
@@ -0,0 +1,239 @@
+  }
+  catch (MooseException & e)
+  {
+    if (((std::string)e.what())
+            .compare("This mesh generator does not work for the provided external boundary as it "
+                     "is not a closed loop.") != 0)
+      paramError("boundary_1", "The provided boundary is not a single-segment boundary.");
```
so if another exception is thrown what happens?

miaoyinb replied:
The boundary is usable to `FillBetweenSidesetsGenerator` only if this exception (invalid_type #2 in previous version) is thrown by `isBoundaryValid()`. The "valid" in `isBoundaryValid()` is for `PeripheralRingMeshGenerator`.

It occurred to me that another error message is needed when isBoundaryValid()'s return is `true`. Added.

(Maybe I need to change the method name to `isBoundaryClosedLoop()`?

miaoyinb replied:
Renamed as `isBoundarySimpleClosedLoop()` in the latest commit. Now I think it makes more sense here.

GiudGiud replied:
umm good point.
My concern was that if the routine throws with a different exception, then the if is not triggered and we just move forward with no error

GiudGiud replied:
I think that is still the case?

miaoyinb replied:
In the latest commit (see Lines 154 thru 170), there are following four possible cases here:

(1) isBoundarySimpleClosedLoop() returns `true`:  This means that the boundary is a closed loop, which is not supported by `FillBetweenSideSetsGenerator`. Then `paramError()`  in the `try` bracket is triggered and throw **"The provided boundary is closed loop, which is not supported."**

If isBoundarySimpleClosedLoop() returns `false`, one of the three exception messages is thrown:
(2) "This mesh generator does not work for the provided external boundary as it has more than one segments." The `paramError()` in the `catch` bracket is triggered and throw **"The provided boundary is not an open single-segment boundary."** 

(3) "This mesh generator does not work for the provided external boundary as it is not a closed loop." This is the only case in which `FillBetweenSideSetsGenerator` works (i.e., an **open and one-segment boundary**). It will pass the `catch` bracket without triggering `paramError()`

(4) "This mesh generator does not work for the provided external boundary as azimuthal angles of consecutive nodes do not change monotonically." The `paramError()` in the `catch` bracket is triggered and throw **"The provided boundary is not an open single-segment boundary."** 

I think this covers all the scenarios.

GiudGiud replied:
ok right now it covers all the scenarios. But these tools will change. I think you should have a generic catcher for when an exception is thrown that has not been hardcoded in the downstream object. 
It's an else to the if that you added

miaoyinb replied:
Here the only `else` is that the exception message "This mesh generator does not work for the provided external boundary as it is not a closed loop." is thrown. In that case, the `else` is what `FillBetweenSideSetsGenerator` wants.

But I do understand your concern. I think I will move the `try` and `catch` here into a new function called `isOpenSingleSegmentBoundary()` in the tools namespace to eliminate confusion here.

GiudGiud replied:
but then you will need to pass name/type to the toolbox to make a clear error message.

I think the 
```
try
{
}
catch
{
  if (this known exception)
    this_message
  else
    return this other exception message that wasnt planned when the class was created
}
```

is fine

miaoyinb replied:
I used another set of moose exceptions to prevent name/type passing. See the commit and let me know your comment. Thanks.

miaoyinb committed:
improve error messages ref #20641

rename isValidBoundary ref #20641

clang-format fix

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1031575/) on e8befe1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20645/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20645/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 62e10e5060452fb79910b76f6d9fd29d406a921b`

miaoyinb committed:
sidesets ids consistency ref. #20641

additional mod

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1032251/) on 1e5cf50 : invalidated by @miaoyinb

moosebuild commented:
Job [End user Apple Silicon](https://civet.inl.gov/job/1032671/) on 0f2fefd : invalidated by @miaoyinb

miaoyinb committed:
better exceptions management ref. #20641

GiudGiud commented on framework/src/utils/FillBetweenPointVectorsTools.C:
```diff
@@ -0,0 +1,655 @@
+  {
+    isBoundarySimpleClosedLoop(mesh, max_node_radius, boundary_ordered_node_list, origin_pt, bid);
+  }
+  catch (MooseException & e)
+  {
+    if (((std::string)e.what())
+            .compare("This mesh generator does not work for the provided external boundary as it "
+                     "is not a closed loop.") != 0)
+    {
+      throw MooseException("The provided boundary is not an open single-segment boundary.");
+      return false;
+    }
+    else
+      return true;
+  }
```
other instances are good to go but this one still seems like it wont age well with new exceptions being added to the called routines. 

If you get an exception, and it s not the one you expect, then you return all good / true.

GiudGiud commented on framework/src/utils/FillBetweenPointVectorsTools.C:
```diff
@@ -0,0 +1,655 @@
+    else
+      return true;
+  }
+  throw MooseException("The provided boundary is closed loop, which is not supported.");
+  return false;
```
this is on the not caught path though? Why an exception?

GiudGiud replied:
I must be misunderstanding something here

miaoyinb replied:
No exceptions means the boundary is a closed loop, which is not an open single segment. Therefore, an exception is thrown and return value is `false` for this function.

GiudGiud reviewed (changes requested):
not sure about this last exception catching

miaoyinb commented:
> not sure about this last exception catching

I think it catches everything now. Let me know your idea. Thanks.

GiudGiud commented:
Thanks for the contribution! Will definitely come in handy for further mesh generation efforts

miaoyinb commented:
Thank you so much for your review, @GiudGiud @cticenhour
